velox/experimental/cudf/exec/VeloxCudfInterop.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/experimental/cudf/exec/NvtxHelper.h"
+#include "velox/experimental/cudf/exec/Utilities.h"
+#include "velox/experimental/cudf/exec/VeloxCudfInterop.h"
+
+#include "velox/common/memory/Memory.h"
+#include "velox/type/Type.h"
+#include "velox/vector/BaseVector.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DictionaryVector.h"
+#include "velox/vector/FlatVector.h"
+#include "velox/vector/arrow/Bridge.h"
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/interop.hpp>
+#include <cudf/strings/string_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <nvtx3/nvtx3.hpp>
+#include <thrust/copy.h>
+#include <thrust/execution_policy.h>
+
+#include <arrow/c/bridge.h>
+#include <arrow/io/interfaces.h>
+#include <arrow/table.h>
+
+#include <functional>
+#include <numeric>
+
+namespace facebook::velox::cudf_velox {
+
+namespace with_arrow {
+
+std::unique_ptr<cudf::table> toCudfTable(
+    const facebook::velox::RowVectorPtr& veloxTable,
+    facebook::velox::memory::MemoryPool* pool,
+    rmm::cuda_stream_view stream) {
+  // Need to flattenDictionary and flattenConstant, otherwise we observe issues
+  // in the null mask.
+  ArrowOptions arrowOptions{true, true};
+  ArrowArray arrowArray;
+  exportToArrow(
+      std::dynamic_pointer_cast<facebook::velox::BaseVector>(veloxTable),
+      arrowArray,
+      pool,
+      arrowOptions);
+  ArrowSchema arrowSchema;
+  exportToArrow(
+      std::dynamic_pointer_cast<facebook::velox::BaseVector>(veloxTable),
+      arrowSchema,
+      arrowOptions);
+  auto tbl = cudf::from_arrow(&arrowSchema, &arrowArray, stream);
+
+  // Release Arrow resources
+  if (arrowArray.release) {
+    arrowArray.release(&arrowArray);
+  }
+  if (arrowSchema.release) {
+    arrowSchema.release(&arrowSchema);
+  }
+  return tbl;
+}
+
+namespace {
+
+void toSignedIntFormat(char* format) {
+  VELOX_CHECK_NOT_NULL(format);
```

Comment:
Can you describe the data type supported in PR description?

Reply:
All non complex velox types should be supported. What you see here is just dictionary indices' type being converted because of difference between cudf and velox.

Reply:
Arrow does not supports the null type which is `Unknown` type in velox. What if the RowVector contains complex data type? Maybe add the data type check for fallback to CPU operator.